highlighters/main/main-highlighter.zsh

```diff
@@ -196,7 +196,21 @@ _zsh_highlight_main__stack_pop() {
 _zsh_highlight_highlighter_main_paint()
 {
   # Before we even 'emulate -L', save the user's options
-  local -A useroptions; useroptions=("${(@kv)options}")
+  local -A useroptions
+  if zmodload -e zsh/parameter; then
+    useroptions=("${(@kv)options}")
+  else
+    local canonicaloptions onoff option rawoptions
```

Comment:
Suggest to wrap the contents of the `else` block into an anonymous function so these variables don't persist for the rest of `_zsh_highlight_highlighter_main_paint`.  The anonymous function will set $useroptions which will have been declared by its caller.

Reply:
I'm not sure it's worth it. If there is an error, the message would start `(anon)` rather than the more helpful `_zsh_highlight_highlighter_main_paint`.

Reply:
This should be fixed in zsh itself.

A block isn't a scope, right? `{ local a; } echo $+a` so either an explicit `unset` or punt. Your call.